GeoLib/PointVec.cpp

```diff
@@ -85,102 +117,40 @@ void PointVec::push_back (Point* pnt, std::string const*const name)
-
-	return _data_vec->size()-1;
-}
-
-void PointVec::makePntsUnique (std::vector<GeoLib::Point*>* pnt_vec,
-                               std::vector<std::size_t> &pnt_id_map, double eps)
+std::size_t PointVec::uniqueInsert(Point* pnt)
 {
-	std::size_t const n_pnts_in_file(pnt_vec->size());
-	std::vector<std::size_t> perm(n_pnts_in_file);
-	pnt_id_map.resize(n_pnts_in_file);
-	std::iota(perm.begin(), perm.end(), 0);
-	std::iota(pnt_id_map.begin(), pnt_id_map.end(), 0);
-
-	// sort the points
-	BaseLib::Quicksort<GeoLib::Point*>(*pnt_vec, 0, n_pnts_in_file, perm);
-
-	// unfortunately quicksort is not stable -
-	// sort identical points by id - to make sorting stable
-	// determine intervals with identical points to resort for stability of sorting
-	std::vector<std::size_t> identical_pnts_interval;
-	bool identical(false);
-	for (std::size_t k = 0; k < n_pnts_in_file - 1; k++)
-	{
-		if (MathLib::maxNormDist((*pnt_vec)[k + 1], (*pnt_vec)[k]) <= eps)
-		{
-			// points are identical, sort by id
-			if (!identical)
-				identical_pnts_interval.push_back(k);
-			identical = true;
-		} else {
-			if (identical)
-				identical_pnts_interval.push_back(k + 1);
-			identical = false;
-		}
-	}
-	if (identical)
-		identical_pnts_interval.push_back(n_pnts_in_file);
-
-	for (std::size_t i(0); i < identical_pnts_interval.size() / 2; i++) {
-		// bubble sort by id
-		std::size_t beg(identical_pnts_interval[2 * i]);
-		std::size_t end(identical_pnts_interval[2 * i + 1]);
-		for (std::size_t j(beg); j < end; j++)
-			for (std::size_t k(beg); k < end - 1; k++)
-				if (perm[k] > perm[k + 1])
-					std::swap(perm[k], perm[k + 1]);
+	GeoLib::Point * ret_pnt(nullptr);
+	if (_oct_tree->addPoint(pnt, ret_pnt)) {
+		_data_vec->push_back(pnt);
+		return _data_vec->size()-1;
 	}
 
-	// check if there are identical points
-	for (std::size_t k = 0; k < n_pnts_in_file - 1; k++)
-		if (MathLib::maxNormDist((*pnt_vec)[k + 1], (*pnt_vec)[k]) <= eps)
-			pnt_id_map[perm[k + 1]] = pnt_id_map[perm[k]];
-
-	// reverse permutation
-	BaseLib::Quicksort<std::size_t, GeoLib::Point*>(perm, 0, n_pnts_in_file, *pnt_vec);
-
-	// remove the second, third, ... occurrence from vector
-	for (std::size_t k(0); k < n_pnts_in_file; k++) {
-		if (pnt_id_map[k] < k) {
-			delete (*pnt_vec)[k];
-			(*pnt_vec)[k] = nullptr;
+	if (ret_pnt == nullptr) { // pnt is outside of OctTree object
+		// update the axis aligned bounding box
+		_aabb.update(*pnt);
+		// recreate the (enlarged) OctTree
+		_oct_tree.reset(GeoLib::OctTree<GeoLib::Point, 16>::createOctTree(
+			_aabb.getMinPoint(), _aabb.getMaxPoint(), _rel_eps));
+		// add all points that are already in the _data_vec
+		for (std::size_t k(0); k<_data_vec->size(); ++k) {
+			GeoLib::Point *const p((*_data_vec)[k]);
+			_oct_tree->addPoint(p, ret_pnt);
 		}
+		// add the new point
+		ret_pnt = nullptr;
+		_oct_tree->addPoint(pnt, ret_pnt);
+		pnt->setID(_data_vec->size());
+		_data_vec->push_back(pnt);
+		return _data_vec->size()-1;
```

Comment:
This can not happen since in line 143 something is pushed into the vector `_data_vec`.